content/post/hercules.v4.md

```diff
@@ -0,0 +1,295 @@
+corresponding cloned items are garbage collected.
+
+When the whole history is processed Hercules asks "leaf" pipeline items to
+generate the result. Those results are serialized and written to the output stream.
+It is also possible to load several serialized results from disk and combine
+them together - for example, to obtain the grand view across the company's repositories.
+
+This versatile architecture allows Hercules to efficiently analyze small and big
+Git repositories while minimizing the complexity and boilerplate code to add
+new analysis types.
+
+# v4 challenges
+
+There were three most difficult technical challenges which had to solved in v3 and v4:
+
+1. Merging results together.
+2. Forks and merges behavior of `PipelineItem`-s.
+3. Determine the order in which to process commits, forks, merges and garbage collect branches.
+
+The first two were especially hard to solve for the line burndown analysis. Regarding (1),
+the results can be sampled and band-split with different frequencies, so it was
+required to resample and interpolate the matrices by day, sum them and then
+sample and split into bands again.
+[One of the scariest functions I have ever written.](https://github.com/src-d/hercules/blob/99abfb229a880d55462eecd2373433df686f8fcc/leaves/burndown.go#L441).
+Regarding (2), it was needed to write the code to merge multiple line interval-marked
+files together and organize the work with a partially shared, partially copied state.
+Git merge does not necessarily contain two branches: it can be three, four, etc.
+
+{{% caption src="/post/hercules.v4/git_abstract.png" %}}
+Abstract from [git/git](https://github.com/git/git) history.
+{{% /caption %}}
+
+I did not come up with a solution better than mark the changed line intervals in files
+while consuming a merge commit with a special outlier value, then
+annotate each line of the each merged file, and copy the only real last edit date
+over those marked as outliers. In places of same edits but conflicting dates we
+choose the oldest date. In places of merge conflicts we automatically get the date
+of the merge commit without special tricks. We should not compare the files
+which are not involved in the merge, so we keep track of their hashes, we maintain
+them. I also had to rewrite the statistics storage logic which used to be too
+"sequential". Overall, fulfilling (2) was a technical nightmare.
+
+Surprisingly, however, most of the efforts went to (3). The first problem was
+related to minimizing the amount of forks and merges, in particular removing
+the no-op fast-forward-like back edges of the DAG.
+
+{{% caption src="/post/hercules.v4/ff.gif" %}}
+Example of the removal of a fast-forward DAG edge.
+{{% /caption %}}
+
+We [topologically sort](https://en.wikipedia.org/wiki/Topological_sorting)
+the nodes of our DAG and traverse it starting from the root.
+Each time we suspect that an edge can be removed we traverse backwards from the children
```

Comment:
I may be wrong, but American English seems to use a different spelling `traverse backwards` -> `traverse backward`. Because reasons...